### PR TITLE
Set node alias

### PR DESCRIPTION
**Pull request by enigbe**
## What this PR does:
- Adds a function to allow users set the node alias.
- If set, ensures that the provided alias is broadcast with the node announcement

## Related issue(s):
- #304

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -132,6 +132,8 @@ pub enum BuildError {
 	WalletSetupFailed,
 	/// We failed to setup the logger.
 	LoggerSetupFailed,
+	/// The provided alias is invalid
+	InvalidNodeAlias(String),
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/builder.rs]: I have simplified the error variant. Initially I wanted to provide context to the caller so they can be informed about why the provided alias is being rejected, but I ran into problems with the language bindings.

File: src/builder.rs [hunk 2]
```diff
@@ -302,6 +307,17 @@ impl NodeBuilder {
 		self
 	}
 
+	/// Sets the alias the [`Node`] will use in its announcement. The provided
+	/// alias must be a valid UTF-8 string.
+	pub fn set_node_alias<T: Into<String>>(
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/builder.rs]: Done.

File: src/builder.rs [hunk 3]
```diff
@@ -1043,3 +1059,68 @@ fn seed_bytes_from_config(
...
+	let node_alias: String = node_alias.into();
+	let alias = node_alias.trim();
+
+	// Alias is non-empty
+	if alias.is_empty() {
+		return Err(BuildError::InvalidNodeAlias("Node alias cannot be empty.".to_string()));
+	}
+
+	// Alias valid up to first null byte
+	let first_null = alias.as_bytes().iter().position(|b| *b == 0).unwrap_or(alias.len());
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/builder.rs]: I took a cursory look at rust-lightning and borrowed [part of their validation approach](https://github.com/lightningdevkit/rust-lightning/blob/3d76753a10a0a44a5598164b251c3ce105fb45bb/lightning/src/routing/gossip.rs#L1399) before displaying `NodeAlias`es. I thought it would be a useful validation at this level rather than allowing aliases with sandwiched `null` bytes to get to the protocol level.

File: src/lib.rs [hunk 4]
```diff
@@ -652,7 +653,16 @@ impl Node {
 								continue;
 							}
 
-							bcast_pm.broadcast_node_announcement([0; 3], [0; 32], addresses);
+							// Extract alias if set, else select the default
+							let alias = if let Some(ref alias) = node_alias {
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/lib.rs]: Done.

**Comment by enigbe** [comment]
> Thanks! I think beyond checking for the node alias and listening addresses before broadcasting, we should also modify the `default_user_config` method in `Config` to, if any of the two are unconfigured:
> 
> a) set [`UserConfig::accept_forwards_to_priv_channels`](https://docs.rs/lightning/latest/lightning/util/config/struct.UserConfig.html#structfield.accept_forwards_to_priv_channels) to `false` b) set [`ChannelHandshakeConfig::announced_channel`](https://docs.rs/lightning/latest/lightning/util/config/struct.ChannelHandshakeConfig.html#structfield.announced_channel) to `false` c) set [`ChannelHandshakeLimits::force_announced_channel_preference`](https://docs.rs/lightning/latest/lightning/util/config/struct.ChannelHandshakeLimits.html#structfield.force_announced_channel_preference) to `true`
> 
> Additionally, we'll want to return an error from `connect_open_channel` if any of the two are unconfigured and we try to open an announced channel. To this end, it might be bit cleaner if we'd split the method in two (`open_channel` and `open_announced_channel`) rather than having it take a simple `bool`.
> 
> TLDR: We should use the configured listening addresses and node alias as indicators whether we are a forwarding node or not. If we're not, we should disallow announced channels.
> 
> Let me know if you'd be up for these additional changes, otherwise I'm happy to pick them up as a follow-up.

Thanks for the review. I am happy to keep working on this until it is ready.

**Comment by enigbe** [comment]
> > Thanks for the review. I am happy to keep working on this until it is ready.
> 
> Thanks again! Do you still intend to do this in this PR, or would you prefer we land this and do additional changes in a follow up?
> 
> Btw. this also needs a rebase already, and there will be larger code changes upcoming in the LDK/BDK/rust-bitcoin upgrade PR, which will likely also lead to further rebase conflicts.

Apologies for the delay. I am almost done and will be pushing changes today. I'll rebase and get it ready for another look before the end of the day.

File: src/builder.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (enigbe) [reply src/builder.rs]: This has been removed and is not relevant.

File: README.md [hunk 5]
```diff
@@ -60,6 +60,9 @@ LDK Node currently comes with a decidedly opinionated set of design choices:
 - Gossip data may be sourced via Lightning's peer-to-peer network or the [Rapid Gossip Sync](https://docs.rs/lightning-rapid-gossip-sync/*/lightning_rapid_gossip_sync/) protocol.
 - Entropy for the Lightning and on-chain wallets may be sourced from raw bytes or a [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) mnemonic. In addition, LDK Node offers the means to generate and persist the entropy bytes to disk.
 
+**Note**: 
+Regarding node announcements, we have decided not to broadcast these announcements if either the node's listening addresses or its node alias are not set.
```
[the remark this answers is not in the document]
Comment (enigbe) [reply README.md]: I agree too. I wasn't sure where it belonged but this [comment](https://github.com/lightningdevkit/ldk-node/pull/330#discussion_r1706684829) gave me a hint.

File: src/lib.rs [hunk 6]
```diff
@@ -603,6 +603,7 @@ impl Node {
 		let bcast_logger = Arc::clone(&self.logger);
 		let bcast_ann_timestamp = Arc::clone(&self.latest_node_announcement_broadcast_timestamp);
 		let mut stop_bcast = self.stop_sender.subscribe();
+		let node_alias = self.config().node_alias;
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/lib.rs]: Done.

File: src/lib.rs [hunk 7]
```diff
@@ -646,13 +647,18 @@ impl Node {
 							}
 
 							let addresses = bcast_config.listening_addresses.clone().unwrap_or(Vec::new());
+							let alias = node_alias.clone().map(|alias| {
+								let mut buf = [0_u8; 32];
+								buf[..alias.as_bytes().len()].copy_from_slice(alias.as_bytes());
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/lib.rs]: I do agree with this approach. At the time I submitted the PR I was still getting familiar with the codebase and didn't know about the custom type converter. I have implemented it for `NodeAlias` with the necessary sanitization to ensure that both methods of setting the node alias are safe.

**Comment by enigbe** [comment]
> > > > Thanks for the review. I am happy to keep working on this until it is ready.
> > > 
> > > 
> > > Thanks again! Do you still intend to do this in this PR, or would you prefer we land this and do additional changes in a follow up?
> > > Btw. this also needs a rebase already, and there will be larger code changes upcoming in the LDK/BDK/rust-bitcoin upgrade PR, which will likely also lead to further rebase conflicts.
> > 
> > 
> > Apologies for the delay. I am almost done and will be pushing changes today. I'll rebase and get it ready for another look before the end of the day.
> 
> Hi @tnull, I have rebased and pushed recommended changes. Unfortunately, the integration tests are flaky and I am uncertain about why. Currently investigating.

Hi @tnull, thanks for your review so far. Have you had some time to check out the recent changes I have made? I have addressed the comments raised and would be grateful for any further feedback or recommendation.

File: src/lib.rs [hunk 8]
```diff
@@ -1250,12 +1261,13 @@ impl Node {
 		}
 
 		let mut user_config = default_user_config(&self.config);
-		user_config.channel_handshake_config.announced_channel = announce_channel;
+		let can_announce_channel = can_announce_channel(&self.config);
+		user_config.channel_handshake_config.announced_channel = can_announce_channel;
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/lib.rs]: Done the reversal as advised.

File: src/error.rs [hunk 9]
```diff
@@ -101,6 +103,10 @@ pub enum Error {
 	LiquiditySourceUnavailable,
 	/// The given operation failed due to the LSP's required opening fee being too high.
 	LiquidityFeeTooHigh,
+	/// Returned when trying to open an announced channel with a peer. This
+	/// error occurs when a [`crate::Node`'s] alias or listening addresses
+	/// are unconfigured.
+	OpenAnnouncedChannelFailed,
```
[the remark this answers is not in the document]
Comment (enigbe) [reply src/error.rs]: I decided to use the existing error variants as suggested, i.e. `InvalidNodeAlias` and `InvalidSocketAddress`. To ensure that we are selecting the correct variant and that we properly inform callers about why opening a channel might fail, I introduced two enums (see snippet below) so that we can attach more context to channel announcement.

https://github.com/enigbe/ldk-node/blob/4fd1cb8ed98c5875627e31d9c7cbf3c5e0fe8d6b/src/config.rs#L279-L297

https://github.com/enigbe/ldk-node/blob/4fd1cb8ed98c5875627e31d9c7cbf3c5e0fe8d6b/src/config.rs#L304C1-L320C2

**Comment by enigbe** [comment]
> Seems CI is failing. Let me know when this is ready for another round of review!

I have made some changes to address flaky tests that were failing in CI. They are all passing now with the exception of `Rust Tests / build(windows-latest, 1.63.0)` which runs endlessly until it is cancelled. It appears stuck building the project. 

I am currently investigating it. I would be grateful for another review while I try to fix the build.

**Comment by enigbe** [comment]
Thank you @tnull for your patient review of this and for cleaning up the nits in #366.

**Comment by enigbe** [comment]
No, I don't mind. I am very grateful. 
I am looking to pick up more work based on issues we have on 0.5 and would be happy/open to be assigned any task.
